cpp/src/arrow/compute/exec/union_node_test.cc

```diff
@@ -0,0 +1,203 @@
+
+  Declaration union_decl{"union", ExecNodeOptions{}};
+
+  // add left source
+  union_decl.inputs.emplace_back(Declaration{
+      "source", SourceNodeOptions{l_batches.schema, l_batches.gen(parallel,
+                                                                  /*slow=*/false)}});
+  // add right source
+  union_decl.inputs.emplace_back(Declaration{
+      "source", SourceNodeOptions{r_batches.schema, r_batches.gen(parallel,
+                                                                  /*slow=*/false)}});
+  AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+
+  ASSERT_OK(Declaration::Sequence({union_decl, {"sink", SinkNodeOptions{&sink_gen}}})
+                .AddToPlan(plan.get()));
+
+  Future<std::vector<ExecBatch>> actual = StartAndCollect(plan.get(), sink_gen);
+
+  auto expected_matcher =
+      Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches)));
+  ASSERT_THAT(actual, expected_matcher);
+}
+
+void RunNonEmptyTest(bool parallel) {
+  auto l_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+  auto r_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+  BatchesWithSchema l_batches, r_batches, exp_batches;
+
+  int multiplicity = parallel ? 100 : 1;
+
+  GenerateBatchesFromString(l_schema,
+                            {
+                                R"([[0,"d"], [1,"b"]])",
+                                R"([[2,"d"], [3,"a"], [4,"a"]])",
+                            },
+                            &l_batches, multiplicity);
+
+  GenerateBatchesFromString(r_schema,
+                            {
+                                R"([[10,"A"]])",
+                            },
+                            &r_batches, multiplicity);
+
+  GenerateBatchesFromString(l_schema,
+                            {
+                                R"([[0,"d"], [1,"b"]])",
+                                R"([[2,"d"], [3,"a"], [4,"a"]])",
+
+                                R"([[10,"A"]])",
+                            },
+                            &exp_batches, multiplicity);
+  CheckRunOutput(l_batches, r_batches, exp_batches, parallel);
+}
+
+void RunEmptyTest(bool parallel) {
+  auto l_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+  auto r_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+
+  int multiplicity = parallel ? 100 : 1;
+
+  BatchesWithSchema l_empty, r_empty, output_batches;
+
+  GenerateBatchesFromString(l_schema, {R"([])"}, &l_empty, multiplicity);
+  GenerateBatchesFromString(r_schema, {R"([])"}, &r_empty, multiplicity);
+
+  GenerateBatchesFromString(l_schema, {R"([])", R"([])"}, &output_batches, multiplicity);
+
+  CheckRunOutput(l_empty, r_empty, output_batches);
+}
+
+TEST(UnionTest, TestNonEmpty) {
+  for (bool parallel : {false, true}) {
+    RunNonEmptyTest(parallel);
+  }
+}
+
+TEST(UnionTest, TestEmpty) {
+  for (bool parallel : {false, true}) {
+    RunEmptyTest(parallel);
+  }
+}
```

Comment:
I think it'd be good to add a test that tries to create a union node with 0 inputs, and with differing schemas, to ensure those cases are covered.